src/gpuarray_array.c

```diff
@@ -988,19 +988,27 @@ int GpuArray_concatenate(GpuArray *r, const GpuArray **as, size_t n,
   }
 
   err = GpuArray_empty(r, as[0]->ops, GpuArray_context(as[0]), restype,
-                       as[0]->nd, dims, GA_ANY_ORDER);
+                       as[0]->nd, dims, GA_C_ORDER);
   free(dims);
   if (err != GA_NO_ERROR) {
     return err;
   }
 
   res_off = r->offset;
   for (i = 0; i < n; i++) {
-    err = r->ops->buffer_extcopy(as[i]->data, as[i]->offset, r->data,
-                                 res_off, as[i]->typecode, r->typecode,
-                                 as[i]->nd, as[i]->dimensions,
-                                 as[i]->strides, r->nd, as[i]->dimensions,
-                                 r->strides);
+    if (axis == 0 &&
+        GpuArray_IS_C_CONTIGUOUS(as[i]) &&
+        as[i]->typecode == r->typecode) {
+      sz = gpuarray_get_elsize(r->typecode);
+      for (j = 0; j < r->nd; j++) sz *= as[i]->dimensions[j];
```

Comment:
I don't see any shapes or nd check anywhere related to concatenate. Why this is missing?

I would more use `as[i]->nd` then `r->nd` in that case. To not read bad data.